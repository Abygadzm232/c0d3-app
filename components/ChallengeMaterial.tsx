import React, { useState } from 'react'
import {
  Challenge,
  ChallengeSubmissionData,
  UserSubmission,
  UserSubmissionsObject
} from '../@types/challenge'

type CurrentChallengeID = string

type ChallengeTitleCardProps = {
  key: string
  id: string
  title: string
  challengeNum: number
  submissionStatus: string
  active?: boolean
  setCurrentChallenge: React.Dispatch<CurrentChallengeID>
}

type ChallengeQuestionCardProps = {
  currentChallenge: Challenge
}

type ChallengeMaterialProps = {
  userSubmissions: UserSubmission[]
  challenges: Challenge[]
}

export const ChallengeTitleCard: React.FC<ChallengeTitleCardProps> = ({
  submissionStatus,
  active,
  title,
  id,
  challengeNum,
  setCurrentChallenge
}) => {
  const cardStyles: string[] = []
  if (active) {
    cardStyles.push('challenge-title-card--active')
  }
  if (submissionStatus === 'passed') {
    cardStyles.push('challenge-title-card--done')
  } else {
    cardStyles.push('shadow-sm', 'border-0', 'challenge-title-card')
  }

  return (
    <div
      data-testid="challenge-title"
      className={`card mb-2 ${cardStyles.join(' ')}`}
      onClick={() => setCurrentChallenge(id)}
    >
      <div className="card-body d-flex justify-content-between">
        <div>{`${challengeNum}. ${title}`}</div>
        {submissionStatus === 'passed' && (
          <img
            width="25px"
            height="25px"
            src="/curriculumAssets/icons/checkmark.svg"
          />
        )}
        {submissionStatus === 'needMoreWork' && (
          <img
            width="25px"
            height="25px"
            src="/curriculumAssets/icons/pending.svg"
          />
        )}
      </div>
    </div>
  )
}

export const ChallengeQuestionCard: React.FC<ChallengeQuestionCardProps> = ({
  currentChallenge
}) => {
  return (
    <div className="card shadow-sm border-0">
      <div className="card-body challenge-question">
        <h1 data-testid="challenge-question-title" className="card-title">
          {currentChallenge.title}
        </h1>
        <div>
          <p
            data-testid="challenge-question-description"
            className="card-question bg-light p-3 mt-3"
          >
            {currentChallenge.description}
          </p>
        </div>
      </div>
    </div>
  )
}

const ChallengeMaterial: React.FC<ChallengeMaterialProps> = ({
  userSubmissions,
  challenges
}) => {
  if (!challenges.length) {
    return <h1>No Challenges for this lesson</h1>
  }
  //create an object to evaluate the student's status with a challenge
  const userSubmissionsObject = userSubmissions.reduce(
    (acc: UserSubmissionsObject, submission: UserSubmission) => {
      acc[submission.challengeId] = submission
      return acc
    },
    {}
  )
  //create a new Challenges array with user submission data integrated
  const challengesWithSubmissionData = challenges.reduce(
    (acc: ChallengeSubmissionData[], challenge: Challenge) => {
      const submissionData = userSubmissionsObject[challenge.id] || {}
      acc.push(
        Object.assign(challenge, {
          status: submissionData.status || 'unsubmitted',
          submissionData
        })
      )
      return acc
    },
    []
  )
  //sort challenges by order
  const sortedChallenges: ChallengeSubmissionData[] = challengesWithSubmissionData.sort(
    (a, b) => a.order - b.order
  )

  //find first challenge that is not passed
  const firstChallenge = sortedChallenges.find(
    (challenge: ChallengeSubmissionData) => challenge.status !== 'passed'
  ) as ChallengeSubmissionData

  const [currentChallengeID, setCurrentChallenge] = useState<
    CurrentChallengeID
  >(firstChallenge.id)
  const currentChallenge = sortedChallenges.find(
    (challenge: ChallengeSubmissionData) => challenge.id === currentChallengeID
  ) as ChallengeSubmissionData
  const challengeTitleCards: React.ReactElement[] = sortedChallenges.map(
    challenge => {
      return (
        <ChallengeTitleCard
          key={challenge.id}
          id={challenge.id}
          challengeNum={challenge.order}
          title={challenge.title}
          setCurrentChallenge={setCurrentChallenge}
          active={challenge.id === currentChallenge.id}
          submissionStatus={challenge.status}
        />
      )
    }
  )
  return (
    <div className="row challenge-display">
      <div className="col-4">{challengeTitleCards}</div>
      <div className="col-8">
        <ChallengeQuestionCard currentChallenge={currentChallenge} />
      </div>
    </div>
  )
}

export default ChallengeMaterial
