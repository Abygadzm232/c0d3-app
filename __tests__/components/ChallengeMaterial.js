import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import ChallengeMaterial from '../../components/ChallengeMaterial'

describe('Curriculum challenge page', () => {
  test('Should render appropriately when no challenges are passed to component', async () => {
    const { container } = render(
      <ChallengeMaterial challenges={[]} userSubmissions={[]} />
    )
    await wait(() => {
      expect(container).toMatchSnapshot()
    })
  })
  const challenges = [
    {
      id: '105',
      description:
        'Write a function that takes in a number and returns true if that number is greater than 5. Otherwise, return false.',
      title: 'Greater than 5',
      order: 0,
      __typename: 'Challenge'
    },
    {
      id: '107',
      description:
        'Write a function that takes in 2 numbers and returns their sum.',
      title: 'Sum of 2 Numbers',
      order: 1,
      __typename: 'Challenge'
    }
  ]
  const props = {
    challenges,
    userSubmissions: []
  }
  const propsWithSubmissions = {
    challenges,
    userSubmissions: [
      {
        id: '3500',
        status: 'open',
        mrUrl: 'github.com/testmrurl',
        diff: 'test diff1',
        viewCount: 0,
        comment: null,
        challengeId: '105',
        reviewer: null,
        createdAt: '1586907809223',
        updatedAt: '1586907825090'
      },
      {
        id: '3501',
        status: 'needMoreWork',
        mrUrl: 'github.com/testmrurl2',
        diff: 'test diff2',
        viewCount: 0,
        comment: 'test comment',
        challengeId: '107',
        reviewer: {
          id: '1',
          username: 'dan'
        },
        createdAt: '1586907809223',
        updatedAt: '1586907825090'
      }
    ]
  }
  test('Should render first challenge by default when user has no submissions', async () => {
    const { container } = render(<ChallengeMaterial {...props} />)
    expect(container).toMatchSnapshot()
  })
  test('Should render first challenge that is not passed when user has submissions', async () => {
    const { container } = render(
      <ChallengeMaterial {...propsWithSubmissions} />
    )
    expect(container).toMatchSnapshot()
  })

  test('Should render clicked challenge within challenge question', async () => {
    const { getAllByTestId, container } = render(
      <ChallengeMaterial {...props} />
    )
    const challengeTitleCard = getAllByTestId('challenge-title')[1]
    await wait(() => {
      fireEvent.click(challengeTitleCard), expect(container).toMatchSnapshot()
    })
  })
})
