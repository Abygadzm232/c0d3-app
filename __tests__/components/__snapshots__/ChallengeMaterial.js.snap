// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Curriculum challenge page Should render appropriately when no challenges are passed to component 1`] = `
<div>
  <h1>
    No Challenges for this lesson
  </h1>
</div>
`;

exports[`Curriculum challenge page Should render clicked challenge within challenge question 1`] = `
<div>
  <div
    class="row challenge-display"
  >
    <div
      class="col-4"
    >
      <div
        class="card mb-2 challenge-title-card shadow-sm border-0"
        data-testid="challenge-title"
      >
        <div
          class="card-body d-flex justify-content-between"
        >
          <div>
            0. Greater than 5
          </div>
        </div>
      </div>
      <div
        class="card mb-2 challenge-title-card challenge-title-card--active shadow-sm border-0"
        data-testid="challenge-title"
      >
        <div
          class="card-body d-flex justify-content-between"
        >
          <div>
            1. Sum of 2 Numbers
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-8"
    >
      <div
        class="card shadow-sm border-0"
      >
        <div
          class="card-body challenge-question"
        >
          <h1
            class="card-title"
            data-testid="challenge-question-title"
          >
            Sum of 2 Numbers
          </h1>
          <div
            class="bg-light p-3 mt-3"
            data-testid="challenge-question-description"
          >
            <span>
              Write a function that takes in 2 numbers and returns their sum.
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Curriculum challenge page Should render first challenge by default when user has no submissions 1`] = `
<div>
  <div
    class="row challenge-display"
  >
    <div
      class="col-4"
    >
      <div
        class="card mb-2 challenge-title-card challenge-title-card--active shadow-sm border-0"
        data-testid="challenge-title"
      >
        <div
          class="card-body d-flex justify-content-between"
        >
          <div>
            0. Greater than 5
          </div>
        </div>
      </div>
      <div
        class="card mb-2 challenge-title-card shadow-sm border-0"
        data-testid="challenge-title"
      >
        <div
          class="card-body d-flex justify-content-between"
        >
          <div>
            1. Sum of 2 Numbers
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-8"
    >
      <div
        class="card shadow-sm border-0"
      >
        <div
          class="card-body challenge-question"
        >
          <h1
            class="card-title"
            data-testid="challenge-question-title"
          >
            Greater than 5
          </h1>
          <div
            class="bg-light p-3 mt-3"
            data-testid="challenge-question-description"
          >
            <span>
              Write a function that takes in a number and returns true if that number is greater than 5. Otherwise, return false.
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Curriculum challenge page Should render first challenge that is not passed when user has submissions 1`] = `
<div>
  <div
    class="row challenge-display"
  >
    <div
      class="col-4"
    >
      <div
        class="card mb-2 challenge-title-card challenge-title-card--active shadow-sm border-0"
        data-testid="challenge-title"
      >
        <div
          class="card-body d-flex justify-content-between"
        >
          <div>
            0. Greater than 5
          </div>
          <img
            height="25px"
            src="/curriculumAssets/icons/pending.svg"
            width="25px"
          />
        </div>
      </div>
      <div
        class="card mb-2 challenge-title-card shadow-sm border-0"
        data-testid="challenge-title"
      >
        <div
          class="card-body d-flex justify-content-between"
        >
          <div>
            1. Sum of 2 Numbers
          </div>
          <img
            height="25px"
            src="/curriculumAssets/icons/pending.svg"
            width="25px"
          />
        </div>
      </div>
    </div>
    <div
      class="col-8"
    >
      <div
        class="card shadow-sm border-0"
      >
        <div
          class="card-body challenge-question"
        >
          <h1
            class="card-title"
            data-testid="challenge-question-title"
          >
            Greater than 5
          </h1>
          <div
            class="bg-light p-3 mt-3"
            data-testid="challenge-question-description"
          >
            <span>
              Write a function that takes in a number and returns true if that number is greater than 5. Otherwise, return false.
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
